pkg/provider/aws/image.go

```diff
@@ -0,0 +1,166 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package aws
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+)
+
+type ImageInfo struct {
+	ImageID      string
+	CreationDate string
+}
+
+type ByCreationDate []ImageInfo
+
+func (a ByCreationDate) Len() int           { return len(a) }
+func (a ByCreationDate) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a ByCreationDate) Less(i, j int) bool { return a[i].CreationDate > a[j].CreationDate }
+
+func (a *Client) getAMI() error {
+	// If the image ID is already set by the user, return
+	if a.Spec.Image.ImageId != nil {
+		return nil
+	}
+
+	// Work around for AWS types
+	filterName := "name"
+	filterArchitecture := "architecture"
+	filterOwner := "owner-id"
+
+	var filterNameValue []string
+	var filterArchitectureValue []string
+
+	if a.Spec.Instance.Image.Architecture != "" {
+		switch a.Spec.Instance.Image.Architecture {
+		case "x86_64":
+			filterArchitectureValue = []string{"x86_64", "amd64"}
+		case "arm64":
+			filterArchitectureValue = []string{"arm64"}
+		case "amd64":
+			filterArchitectureValue = []string{"x86_64", "amd64"}
+		default:
+			return fmt.Errorf("invalid architecture %s", a.Spec.Instance.Image.Architecture)
+		}
+	}
+
+	// Default to the official Ubuntu images in the AWS Marketplace
+	// TODO: Add support for other image OS types
+	awsOwner := []string{"099720109477", "679593333241"}
+	if a.Spec.Instance.Image.OwnerId != nil {
+		awsOwner = []string{*a.Spec.Instance.Image.OwnerId}
+	}
+
+	for _, arch := range filterArchitectureValue {
+		filterNameValue = append(filterNameValue, fmt.Sprintf("ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-%s-server-%s*", arch, getCurrentYear()))
+	}
+
+	var nextToken *string
+	var images []ImageInfo
+
+	for {
+		out, err := a.ec2.DescribeImages(context.TODO(), &ec2.DescribeImagesInput{
+			NextToken: nextToken,
+			Filters: []types.Filter{
+				{
+					Name:   &filterName,
+					Values: filterNameValue,
+				},
+				{
+					Name:   &filterArchitecture,
+					Values: filterArchitectureValue,
+				},
+				{
+					Name:   &filterOwner,
+					Values: awsOwner,
+				},
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("failed to describe images: %v", err)
+		}
+
+		if len(out.Images) == 0 {
+			return fmt.Errorf("no images found * %v", filterNameValue)
+		}
+
+		for _, image := range out.Images {
+			images = append(images, ImageInfo{
+				ImageID:      *image.ImageId,
+				CreationDate: *image.CreationDate,
+			})
+		}
+
+		if out.NextToken != nil {
+			nextToken = out.NextToken
+		} else {
+			break
+		}
+	}
+
+	if len(images) == 0 {
+		return fmt.Errorf("no images found")
+	}
+	sort.Sort(ByCreationDate(images))
+	a.Spec.Image.ImageId = &images[0].ImageID
+
+	return nil
+}
+
+func (a *Client) imageIdSupported() error {
+	var nextToken *string
+
+	for {
```

Comment:
do we need to factor thsi out into a `describeImages` helper since we do essentially the same thing with different input in two places.

Reply:
done, we now have a `func describeImages`